docs/proposals/component-registration.md

```diff
@@ -0,0 +1,409 @@
+	Status ExternalServiceProviderStatus
+}
+
+type ExternalServiceProviderSpec struct {
+	Host string
+	Ports []ExternalServiceProviderPort
+	LivenessProbe *Probe
+	ReadinessProbe *Probe
+}
+
+type ExternalServiceProviderPort struct {
+	Name string // optional - for mapping from ServicePort
+	Protocol Protocol // TCP, UDP
+	Port int
+}
+
+type ExternalServiceProviderStatus struct {
+	Conditions []ExternalServiceProviderCondition
+}
+
+type ExternalServiceProviderCondition struct {
+	Type ExternalServiceProviderConditionType // Readiness, Liveness
+	Status ConditionStatus // True, False, Unknown
+	LastUpdateTime util.Time
+	LastTransitionTime util.Time
+	Reason string
+	Message string
+}
+
+type ExternalServiceProviderList struct {
+	TypeMeta
+	ListMeta
+	Items []ExternalServiceProvider
+}
+```
+
+```
+type ServiceStatus struct {
+	...
+	Conditions []ServiceCondition
+}
+
+type ServiceCondition struct {
+	Type ServiceConditionType // Readiness, Liveness
+	Status ConditionStatus // True, False, Unknown
+	LastUpdateTime util.Time
+	LastTransitionTime util.Time
+	Reason string
+	Message string
+}
+```
+
+
+### Kubectl (CLI)
+
+Ready Service Endpoints can be queried with the existing `kubectl get endpoints <service-name>`.
+
+Add a `kubectl get externalserviceproviders` (`kubectl get esp` for short) command to list the external service provider instances, their address/port/name and their last known conditions. Components can be retrived with a component specific label, like `type=component` or `component=apiserver`.
+
+
+### Storage
+
+Store ExternalServiceProviders spec/status in etcd, with an apiserver storage wrapper.
+
+
+### Status Updates
+
+Add an external-service-provider-controller that regularly probes registered providers and updates their stored last known status.
+
+Add self-registration to the components that uses the new ExternalServiceProviders API (Create call) when the component starts.
+
+Add self-deregistration to the components that uses the new ExternalServiceProviders API (Remove call) when the component stops gracefully (e.g. after receiving SIGTERM).
+
+Add self-updating to the components that uses the new ExternalServiceProviders API (Update call) when the component errors/panics.
+
+Use a TCP probe (of `/healthz` or `/healthz/ping`) to check liveness, because it indicates a server is listening on the expected host/path, even if its responses are unhealthy.
+
+Update individual component `/healthz` endpoints with sub-resources to allow them to perform as an indicator of readiness, not just liveness. Readiness should include transitively checking the readiness of immediate dependencies (e.g. etcd or other databases).
+
+
+## Example
+
+The following terminal commands are an example of how the `kubectl get externalserviceproviders` command might work.
+
+TODO
+
+## Alternate Naming
+
+ExternalServiceProvider is a bit verbose for a top-level API resource name. Here are some other naming options for the same concept.
```

Comment:
Let me rephrase: I'm trying to make the terminology and style used in the API consistent. I intend to reserve "external" for the purpose I stated. I plan to rename or deprecate `externalID`. Node identification and addressing is sort of a mess (LegacyHostIP, etc.).

I do not want to use the term "Service", either, as that is a concept that is intentionally decoupled from the application implementing the functionality. I'll defer on that to @thockin, though, since he's considering refactoring the distribution of responsibilities between at least Ingress and Service.

That said, names are easy to change, so long as you are willing to do.